### PR TITLE
Fix: Issue #2279 - Implementing the GlobalCopyPropagation pass

### PR DESCRIPTION
**Pull request by filipsyrmia**
This PR introduces a new pass with the idea of implementing copy propagation logic, but on a 'global' level (regarding control blocks), with the goal of optimizing situations similar to the one described in the #2279 issue. The optimization described isn't a responsibility of the StrengthReduction pass' logic (as mentioned in the issue), it seems like it should be covered in the copy propagation logic of this compiler.

The logic was implemented as a standalone pass rather than expanding upon the existing LocalCopyPropagation pass, this was done to avoid needlessly complicating the already complex logic of the LocalCopyPropagation pass. Having said that, the new pass was designed with the idea of being run in tandem with the LocalCopyPropagation pass and doesn't introduce some of the optimizations that are already included in the existing pass. (This may warrant putting these 2 passes in a new pass manager and running them as such, for now I have just placed them as one after another in the 'p4test/midend.cpp', where the GlobalCopyPropagation is run first because I believe this order is more logical when you look at the optimizations it introduces)

The optimization mentioned in the issue #2279 is accomplished by visiting the body of the action after the 'ActionCall' node, that corresponds to the action mentioned, is visited. This way we retroactively edit the body of the action and propagate any values that are known at that time.

The only thing that seems problematic to me, and was during the process of developing this pass, is that doesn't seem to be a way of retroactively editing the body of the action without using the 'const_cast'  feature. This is needed because after resolving the 'ActionCall' in the preorder function of IR node 'MethodCallExpression' the action pointer that is available to use is a 'const IR::P4Action*' and doing any editing to the body of the action pointed to by this pointer is impossible, therefore the solution I found is forcefully getting rid of the problematic 'const' part. 
Any suggestions regarding this problem are welcome.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
I assume this pass works only for actions that are directly called and have exactly 1 call. It won't handle actions that appear in tables. Is that correct?

**Comment by mbudiu-vmw** [comment]
Then you should document this fact: this pass has to be run after LocalizeAllActions, which ensures that each action is invoked only once. If there is an easy way to check this and throw if not, it would be even better.

**Comment by filipsyrmia** [comment]
I looked around in the other passes of the compiler and haven't found a way of doing a check and throw, so I'll add a '@pre' comment to document this dependency.
I'll also add TODO comments for expanding the pass to work on actions inside of tables as well.

Is there anything else that needs to be added/changed?

**Comment by mbudiu-vmw** [comment]
I still have to do an in-depth review, but understanding these preconditions makes the review much easier.
I would not add a TODO necessarily, just specify clearly what the limitations of the pass are.
If you want a TODO you better file an issue.
Frankly I don't think this will apply much to actions in tables.

File: testdata/p4_16_samples_outputs/psa-resubmit-bmv2-midend.p4 [hunk 1]
```diff
@@ -36,7 +36,6 @@ parser IngressParserImpl(packet_in pkt, out headers_t hdr, inout metadata_t user
 
 control cIngress(inout headers_t hdr, inout metadata_t user_meta, in psa_ingress_input_metadata_t istd, inout psa_ingress_output_metadata_t ostd) {
     @noWarnUnused @name(".send_to_port") action send_to_port_0() {
-        ostd.drop = false;
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples_outputs/psa-resubmit-bmv2-midend.p4:39]: why is this removed?
this looks wrong to me.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: midend/global_copyprop.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    if (auto mem = mc->method->to<IR::Member>())
+        return;
+
+    LOG5("Working on 'MethodCallexpression': " << mc);
+    auto *mi = MethodInstance::resolve(mc, refMap, typeMap, true);
+    // Add an entry in the 'actions' map for the action being called.
+    // Remove entries in the 'vars' map for variables that are used as 'Out' or 'InOut' parameters.
+    if (auto aCall = mi->to<ActionCall>()) {
+        LOG6("  Is 'ActionCall'. Adding entry for action: " << aCall->action);
+        (*actions)[aCall->action] = new std::map<cstring, const IR::Expression*>(vars);
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply midend/global_copyprop.cpp:146]: This is called "defensive programming". We have no way to mechanically check that preconditions are satisfied, so if a pass can validate preconditions it's more robust. We frequently reorder passes or add additional invocations of passes.

File: testdata/p4_16_samples_outputs/gauntlet_mux_validity-bmv2-midend.p4 [hunk 3]
```diff
@@ -29,30 +29,20 @@ parser p(packet_in pkt, out Headers hdr, inout Meta m, inout standard_metadata_t
...
-    }
-    @hidden action gauntlet_mux_validitybmv2l37() {
-        b_0 = false;
-    }
-    @hidden table tbl_gauntlet_mux_validitybmv2l37 {
-        actions = {
-            gauntlet_mux_validitybmv2l37();
-        }
-        const default_action = gauntlet_mux_validitybmv2l37();
+        dummy = dummy;
```
[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/gauntlet_mux_validity-bmv2-midend.p4:38]: repeated would probably be too much, but at least once seems to be very beneficial

File: midend/local_copyprop.cpp [hunk 4]
```diff
@@ -661,4 +661,24 @@ IR::ParserState *DoLocalCopyPropagation::postorder(IR::ParserState *state) {
     return state;
 }
 
+// Reset the state of internal data structures after traversing IR,
+// needed for this pass to function correctly when used in a PassRepeated
+IR::P4Program *DoLocalCopyPropagation::postorder(IR::P4Program *prog) {
```
Comment (mbudiu-vmw) [inline midend/local_copyprop.cpp:666]: this can be done in the init_apply function.

File: testdata/p4_16_samples/issue2279.p4 [hunk 5]
```diff
@@ -0,0 +1,24 @@
+header ethernet_t {
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples/issue2279.p4:1]: this is much more readable. thank you.

File: testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4 [hunk 6]
```diff
@@ -37,7 +37,7 @@ control cIngress(inout headers_t hdr, inout metadata_t user_meta, in psa_ingress
         regfile_0.write(32w1, 48w3);
         regfile_0.write(32w2, 48w4);
         tmp = regfile_0.read(32w1);
-        hdr.ethernet.dstAddr = tmp + regfile_0.read(32w2) + 48w281474976710651;
+        hdr.ethernet.dstAddr = regfile_0.read(32w1) + regfile_0.read(32w2) + 48w281474976710651;
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: this is a weird one.
many passes assume that there aren't two operations with side-effects in the same statement.
so this one I would think should not happen

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]
We are very close.
Some of the improvements are very impressive.
The main question is about the side effects.

File: testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4 (hunk 6 above)
Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: The weird thing is that the mechanism used in the LocalCopyPropagation pass (the hasSideEffects class) to detect when a value of a variable has side-effects does not detect that the value of '_tmp_' variable here has any side-effects, this is the reason why it's value does get propagated into the following statement. Since the mechanism used here is the same for all of the passes (I assume the hasSideEffects class is responsible for this throughout the compiler) I don't think there would be any problems with other passes, unless I am missunderstanding something.

The reason this value is not propagated when the pass is run only once is because of the ordering of the statements in the program. Since the ordering be changed arbitrarily, the number of repeated runs of LocalCopyProp pass needed for this kind of behaviour to show can also vary. This is the reason why this situation did not appear before, or at least not consistently enough to be noticed.

If there is still a need to avoid these scenarios, an Inspector can be added in LocalCopyProp which is run on the right side of every AssignmentStatement and sets a flag if it detects that in that right side an ExternMethod exists, and then the LocalCopyPropagation pass can use this flag to determine if it should propagate further variables whose values are also a ExternMethod.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: SideEffects is very simple. We have to figure out why the expression reading the register file is not detected as having side effects. Do you have time to look at this? 

Mihai

Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: It seems to have a noSideEffects annotation attached to it.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: So the thing to do is to understand whether the meaning of this annotation is the same as what the compiler worries about. Let me think about it a little bit.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: This is from the spec:
```
@noSideEffects - Weaker than @pure and describes a function that does not change any hidden state, but may depend on hidden state. One example is a hash function that computes a deterministic hash of its arguments, plus some internal state that can be modified via control plane API calls such as a seed or initialization vector. Another example is a read of one element of a register array extern object. Such a function may be dead code eliminated, and may be reordered or combined with other @noSideEffects or @pure calls (subject to the limits imposed by copy-out behavior of out and inout parameters), but not with other function calls that may have side effects that affect the function. 
```

Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: If I understand correctly, we need to know if other function calls within a statement affect the function call we are about to propagate. This doesn't seem to be checked anywhere in the LocalCopyProp pass, so how would we achieve this?

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: That is a vague statement unfortunately.
The problem is not with this pass, the work done by this pass is correct.
The problem is with other passes, that may expect that there aren't two method calls in the same statement.
It's not easy to review all other passes to make sure that this is what they expect.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: We can leave it like this and see whether anything breaks.
I will approve this for now.

Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: I was wondering, can BuiltInMethods be problematic as well when having multiple in one statement?

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: BTW: this invariant is established by SideEffectOrdering, early in the compiler.
The invariant is that each statement has at most one side effect. 
But the definition of "side-effect" is subtle, as can be seen from the spec, where `@pure` and `@noSideEffect` are two forms of possible non-side effects. So it is not exactly clear what each pass may expect.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: What is different about BuiltInMethods is that the compiler fully understands what they do. But for extern methods the compiler only has a promise provided by their API and these annotations.

**Review by mbudiu-vmw — APPROVED** [review]

File: testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4 (hunk 6 above)
Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: Understood, would you like for me to push the changes for the init_apply function and then merge this or should we skip that as well?

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: That is to some degree a cosmetic change. They are equivalent if you always start the visitor from the P4Program node. But if you invoke the visitor on a subtree, they are not equivalent. If you plan to do it I will wait.

Comment (filipsyrmia) [reply testdata/p4_16_samples_outputs/psa-register-complex-bmv2-midend.p4:40]: I will update the PR with that change and then we can move forward with the merge.
